### PR TITLE
[SPARK-49063][SQL] Fix Between with ScalarSubqueries

### PR DESCRIPTION
**Pull request by mihailomilosevic2001**
### What changes were proposed in this pull request?
Fix for between with ScalarSubqueries.

### Why are the changes needed?
There is a regression introduced from a previous PR https://github.com/apache/spark/pull/44299. This needs to be addressed as between operator was completely broken with resolved ScalarSubqueries.

### Does this PR introduce _any_ user-facing change?
No, the bug is not release yet.

### How was this patch tested?
Tests added to golden file.

### Was this patch authored or co-authored using generative AI tooling?
No.

### CODE REVIEW

File: sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala [hunk 1]
```diff
@@ -4603,6 +4603,15 @@ object SQLConf {
       .booleanConf
       .createWithDefault(true)
 
+  val OPTIMIZE_BETWEEN =
+    buildConf("spark.sql.optimizeBetween")
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala:4607]: It's not only an optimization, but also a correctness fix if the BETWEEN input is nondeterministic. I think a legacy config is better here. How about `spark.sql.legacy.duplicateBetweenInput`? By default it's false, and people can set it to true to restore to the legacy behavior.

**Comment by HyukjinKwon** [comment]
> Fix for between with ScalarSubqueries.

What does this PR fix?

**Comment by mihailom-db** [comment]
@HyukjinKwon This ticket's PR https://issues.apache.org/jira/browse/SPARK-46366 introduced an optimization for Between expression. Unfortunately it lead to ScalarSubqueries being firstly resolved before duplicated. This lead to failure in RewriteCorrelatedScalarSubqueries as resolved Subqueries will not need another alias resolution, so we would end up left joining the same query twice. Left join is unresolved if both sides have the same alias.

**Review by cloud-fan — APPROVED** [review]

**Comment by cloud-fan** [comment]
thanks, merging to master!

**Comment by cloud-fan** [comment]
thanks, merging to master!

**Pull request closed without merging**